spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyManager.java

```diff
@@ -0,0 +1,152 @@
+
+    public LocalApiKeyManager(
+            ApiKeyFetcher apiKeyFetcher,
+            @Value("${number-of-minutes-the-gateway-shall-work-in-offline-mode}") long minutesInOfflineMode) {
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.numberOfMinutesTheGatewayShallWorkInOfflineMode = minutesInOfflineMode;
+    }
+
+    public ApiKey getApiKeyEntityFromLocalCache(String apiKeyValue) {
+        //This is only a fallback solution if connection to admin does not work. Therefore offline mode is triggered.
+        updateGatewayMode(GatewayMode.OFFLINE);
+
+        if (!isLocalApiKeyListInitialized) {
+            throw new LocalApiKeyListWasNotInitializedException("The gateway is currently not able to validate " +
+                    "API keys in offline mode, as the local API cache was not yet initialized. This probably " +
+                    "means that a network connection to CoatRack admin could not yet be established.");
+        } else if (isOfflineWorkingTimeExceeded()) {
+            throw new OfflineWorkingTimeExceedingException("The predefined time for working in offline mode is exceeded. The " +
+                    "gateway will reject every request until a connection to CoatRack admin could be re-established.");
+        } else {
+            return extractApiKeyFromLocalApiKeyList(apiKeyValue);
+        }
+    }
+
+    private boolean isOfflineWorkingTimeExceeded() {
+        return LocalDateTime.now().isAfter(deadlineWhenOfflineModeShallStopWorking);
+    }
+
+    private ApiKey extractApiKeyFromLocalApiKeyList(String apiKeyValue) {
+        log.debug("Trying to extract the API key with the value {} from the local list.", apiKeyValue);
+
+        Optional<ApiKey> optionalApiKey = localApiKeyList.stream().filter(
+                apiKeyFromLocalList -> apiKeyFromLocalList.getKeyValue().equals(apiKeyValue)
+        ).findFirst();
+
+        if (optionalApiKey.isPresent()) {
+            return optionalApiKey.get();
+        } else {
+            throw new ApiKeyNotFoundInLocalApiKeyListException("The API key with the value " + apiKeyValue +
+                    " could not be found in the local API key list.");
+        }
+    }
+
+    @Async
+    @Scheduled(fixedRateString = "${local-api-key-list-update-interval-in-millis}")
+    public void refreshLocalApiKeyCacheWithApiKeysFromAdmin() {
+        log.debug("Trying to update the local API key list by contacting CoatRack admin.");
+
+        try {
+            List<ApiKey> fetchedApiKeyList = apiKeyFetcher.requestLatestApiKeyListFromAdmin();
+            updateApiKeyList(fetchedApiKeyList);
+            updateGatewayMode(GatewayMode.ONLINE);
+        } catch (Exception e) {
+            log.error("API key list fetching process failed: " + e);
+            updateGatewayMode(GatewayMode.OFFLINE);
+        }
+    }
+
+    private void updateApiKeyList(List<ApiKey> fetchedApiKeyList) {
+        Assert.notNull(fetchedApiKeyList, "Fetched API key list was null.");
```

Comment:
Suggestion:
```
if (fetchedApiKeyList == null) throw new InvalidArgumentException("Will not update local API key list, as fetched API key list was null.");
```

Reply:
I kept the ```Assert.notNull()``` part as it is cleaner imo but I adjusted the error message as you suggested.